### PR TITLE
interfaces/builtin/network-setup-{control,observe}: allow using netplan directly

### PR DESCRIPTION
**Pull request by anonymouse64**
This allows snaps to use the version of netplan from the base snap directly from
a shell script instead of needing to bundle netplan inside their own snap.

That in turn allows us to update the spread tests to just have a shell script which
calls netplan from a shell script instead of building netplan into the snap. With
this, we also expand the spread test to ensure that the version of netplan from
all the different base snaps works on all the core releases, as noted in https://github.com/snapcore/snapd/pull/9784#issuecomment-772561519

This should probably have been done as a part of https://github.com/snapcore/snapd/pull/7214, but I don't think we really ever finished that work to look into all the accesses needed by netplan from the base snap to work properly.

This now includes https://github.com/snapcore/snapd/pull/9993 in order to actually
run the netplan spread test on UC20.

### CODE REVIEW

**Review by mvo5 — DISMISSED** [review]
Thanks for this!

**Comment by anonymouse64** [comment]
In the meantime I filed https://github.com/snapcore/snapd/pull/9785 to unbreak master

**Review by pedronis — CHANGES_REQUESTED** [review]

File: tests/lib/snaps/netplan-snap/bin/netplan.sh
Comment (bboozzoo) [inline tests/lib/snaps/netplan-snap/bin/netplan.sh:6]: newline

File: interfaces/builtin/network_setup_observe.go
Comment (anonymouse64) [reply interfaces/builtin/network_setup_observe.go:38]: Which part doesn't make sense to you? All of the accesses added to `-observe` here are for reading or executing the netplan binaries from the base snap. This enables clients like:

```bash
out=$(netplan info)
# parse ip from netplan info output
```

instead of having to bundle all of netplan within their own snap which is awkward and annoying

Comment (pedronis) [reply interfaces/builtin/network_setup_observe.go:38]: I somehow misread the comment apparently

File: interfaces/builtin/network_setup_control.go
Comment (pedronis) [inline interfaces/builtin/network_setup_control.go:38]: it's a bit unclear what to do about that but it's not clear we can mandate for all bases to have neptlan inside them

Comment (anonymouse64) [reply interfaces/builtin/network_setup_control.go:38]: Well, I would expect this only to be used on Ubuntu Core systems, so really this is only a requirement of Ubuntu Core boot bases. We could make the snippet conditional and only applied on core systems if that makes you more comfortable?

Comment (anonymouse64) [reply interfaces/builtin/network_setup_control.go:38]: I understand the concern that the policy allows using the binary from the base snap, even if the base snap does not provide it, but I don't think this should be reason enough to block this pr because:
* there are no boot base snaps out there today that currently don't ship netplan
* the only base snaps that don't ship netplan are not boot base snaps (i.e. the freedesktop and the bare base snaps that I know of)
* if we consider what the effect of allowing a snap to be provided permission to run netplan is when there may eventually be a boot base snap that does not have netplan, the impact is very small, specifically if you built an app today (using this pr) on a base snap that has netplan, then you decide to switch base snaps, then your snap will no longer work, but I don't view this as a problem because we don't automatically move snaps from one base snap to another, and we don't allow base snaps to break the contract of providing the file.
* it's very unclear why someone would expect to be able to use network-setup-control on a non-boot base snap, i.e. what does it mean to be able to run netplan from the base snap when you have a freedesktop base snap? 

Really the concern here is about the interface contract providing access to something that might not be there, which is fine IMHO because there are plenty of other cases where this is true in interfaces such as when there is a resource in /dev/, we provide access via apparmor to the /dev/ file even if it doesn't exist. 

At the end of the day allowing this access will not break snaps and will not permit folks to build a snap that works then someday becomes broken. It only allows folks to knowingly build broken snaps, which they can already do today. Additionally this access addresses a product offering gap which is that in order to actually use netplan you have to essentially maintain your own netplan inside your snap which is a really unfortunate user story.

**Review by jdstrand — APPROVED** [review]
From a security policy point of view, this looks fine. The net_admin denial is annoying, but I suspect many snaps with network-setup-control will also plugs network-control, but be aware that only plugging network-setup-observe will trigger the denial (and net_admin isn't in network-observe). As for running the netplan command from the base where the base might be different than on the host, this is a valid concern that can be mitigated by netplan having a stable API (which is really no different than for other implicit dbus policy with alternate bases). I defer to the snapd team on this point.

File: interfaces/builtin/network_setup_control.go
Comment (pedronis) [reply interfaces/builtin/network_setup_control.go:38]: let's say we consider netplan optional,  the other question is do we expect newer and older versions of netplan client to interoperate with newer/older version of actual system netplan?

I suppose we are ok because in the end it can only use the dbus interface that is the contract here? but actually we let it read the files as well under netplan? is the latter a problem?

**Review by pedronis — COMMENTED** [review]
I think I'm ok with this if we understand the backward/forward compatibility properties of this

File: interfaces/builtin/network_setup_control.go
Comment (anonymouse64) [reply interfaces/builtin/network_setup_control.go:38]: Well there are 2 cases:

1. with an older boot base and a newer app base snap
2. with a newer boot base and an older app base snap

For 1), the only way that the app snap could be broken is if the new netplan binary decides to send a different D-Bus API, which IMHO would be netplan breaking a compatibility guarantee, but I don't know if there exists such a version compatibility guarantee

For 2), the only way that the app snap could be broken is if the new netplan service implementation stops understanding old API, which is 100% a broken compatibility promise, and is the same kind of breakage that would happen without this PR if the client bundles and ships netplan in their app snap anyways (like they must be doing today anyways).

Thinking about this again, I think that if the netplan team says they do not guarantee forward compatibility such that a new netplan binary could some-day send incompatible data to an older netplan service, then we should not do this. I asked the team about this and will report back what I hear.

Regarding reading the files, I don't view this as a problem since really those are just python files underneath /usr/share/netplan/... that the netplan binary needs to read/load in order to work, as netplan doesn't install it's python files inside the typical site-packages directory which is allowed by the python apparmor abstraction AFAICT.

**Comment by anonymouse64** [comment]
Marked as blocked waiting for feedback from netplan team on compatibility guarantees

**Comment by anonymouse64** [comment]
We discussed this a bit after hearing back from the netplan team and they guarantee D-Bus backwards stability (where an old client can talk to a new service and things work as expected), and for forwards compatibility (where a new client talks to an old service), they have added a comment to the source code of the netplan client: https://github.com/CanonicalLtd/netplan/commit/d9d35c606e5dee639610c19ae1ad2a5a493e5d80 wherein they plan on keeping the current behavior where the netplan client will always fallback to using D-Bus when it detects it is in snap confinement. 

As such, we decided that this PR is "ok", as long as we expand spread testing to ensure that the newest netplan from the newest base snap works on the oldest base, that is an app snap calling `netplan apply` with a base of core20 works on Ubuntu Core 16 and 18. 

Marking as unblocked, I will try to update this soonish

File: tests/lib/snaps/netplan-snap/bin/netplan.sh
Comment (pedronis) [reply tests/lib/snaps/netplan-snap/bin/netplan.sh:6]: still the case

File: tests/main/fake-netplan-apply/task.yaml
Comment (pedronis) [inline tests/main/fake-netplan-apply/task.yaml:22]: shouldn't we test all the combinations here as well? or we don't because we do that on core?  anyway testing on classic with matching base seems a bit odd as it's not a typical scenario, if we think the matrix is tested elsewhere we might test here just with the base: core18 one

**Review by pedronis — COMMENTED** [review]
a question about the classic system test

File: tests/main/fake-netplan-apply/task.yaml
Comment (bboozzoo) [reply tests/main/fake-netplan-apply/task.yaml:22]: The core test appears to try all bases on each ubuntu core release, so I agree that it may be useful to just run this one with a single base to save both the execution time and leave less room for random failures.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: tests/lib/snaps/netplan-snap/bin/netplan.sh
Comment (anonymouse64) [reply tests/lib/snaps/netplan-snap/bin/netplan.sh:6]: oops thanks for pointing out

File: tests/main/fake-netplan-apply/task.yaml
Comment (anonymouse64) [reply tests/main/fake-netplan-apply/task.yaml:22]: Yes, I would rather keep the matrix of testing all combos just on UC, since that's where netplan is most likely to be used, and also specifically where our customers are going to use netplan, on Classic ubuntu there are many other options to configure networking, but netplan is supposed to be the first class citizen on core. 

I will update the classic test to use core18 though

Comment (anonymouse64) [reply tests/main/fake-netplan-apply/task.yaml:22]: actually after looking at this again, do we have an issue with the current way it is written to test core on 16.04, core18 on 18.04 and core20 on 20.04? Seems reasonable to me and we don't lose much time having the switch to use a different base snap for each system, so if there's no opposition to testing the version of netplan matching the host OS I would rather keep the classic test as-is

Comment (pedronis) [reply tests/main/fake-netplan-apply/task.yaml:22]: it's longer and gives the impression that things have to match for a reason (maybe is just a matter of commenting regarding the latter)

Comment (anonymouse64) [reply tests/main/fake-netplan-apply/task.yaml:29]: I updated the comment above the `systems` line, please take a look and see if it makes sense, if you still want to just test a single version of the netplan client library, I can do that too of course if really desired

**Review by pedronis — APPROVED** [review]
thanks for the clarification, I didn't remember that one side was mocked in that test

**Pull request merged**
